source/iterator/cdc_iterator.go

```diff
@@ -0,0 +1,258 @@
+		lastModified = time.Unix(0, 0)
+	}
+	cdc := &CDCIterator{
+		userName:         config.UserName,
+		apiToken:         config.APIToken,
+		client:           &http.Client{},
+		lastModifiedTime: lastModified,
+		tomb:             tmbWithCtx,
+		caches:           make(chan []sdk.Record, 1),
+		buffer:           make(chan sdk.Record, 1),
+		ticker:           time.NewTicker(config.PollingPeriod),
+		baseURL:          fmt.Sprintf("https://%s.zendesk.com", config.Domain),
+	}
+
+	cdc.tomb.Go(cdc.startCDC(ctx))
+	cdc.tomb.Go(cdc.flush)
+
+	return cdc, nil
+}
+
+// HasNext return true when buffer is not empty
+func (c *CDCIterator) HasNext(_ context.Context) bool {
+	return len(c.buffer) > 0 || !c.tomb.Alive() // return true in case of go routines dying, error will be returned by Next
+}
+
+// Next will check the case whether to push data into buffer
+func (c *CDCIterator) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case rec := <-c.buffer:
+		return rec, nil
+	case <-c.tomb.Dying():
+		return sdk.Record{}, c.tomb.Err()
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err()
+	}
+}
+
+// startCDC fetches records and set next position with lastmodified time of the ticket
+func (c *CDCIterator) startCDC(ctx context.Context) func() error {
+	return func() error {
+		defer close(c.caches)
+		for {
+			select {
+			case <-c.tomb.Dying():
+				return c.tomb.Err()
+			case <-c.ticker.C:
+				records, err := c.fetchRecords(ctx)
+				if err != nil {
+					return err
+				}
+				if len(records) == 0 {
+					continue
+				}
+				select {
+				case c.caches <- records:
+					pos, err := position.ParsePosition(records[len(records)-1].Position)
+					if err != nil {
+						return err
+					}
+					c.lastModifiedTime = pos.LastModified
+				case <-c.tomb.Dying():
+					return c.tomb.Err()
+				}
+			}
+		}
+	}
+}
+
+func (c *CDCIterator) flush() error {
+	defer close(c.buffer)
+	for {
+		select {
+		case <-c.tomb.Dying():
+			return c.tomb.Err()
+		case cache := <-c.caches:
+			for _, record := range cache {
+				c.buffer <- record
+			}
+		}
+	}
+}
+
+// fetchRecords will export tickets from zendesk api, initial random start_time is set to 0
+func (c *CDCIterator) fetchRecords(ctx context.Context) ([]sdk.Record, error) {
+	if c.nextRun.After(time.Now()) {
+		return nil, nil
+	}
+
+	url := fmt.Sprintf("%s/api/v2/incremental/tickets/cursor.json?start_time=%d", c.baseURL, c.lastModifiedTime.Add(time.Second).Unix()) // add one extra second, to get newer updates only
+
+	// if after URL is available, use that
+	if c.afterURL != "" {
+		url = c.afterURL
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not access the zendesk: %w", err)
+	}
+	req.Header.Add("Authorization", "Basic "+basicAuth(c.userName, c.apiToken))
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("could not get the zendesk response: %w", err)
+	}
+	defer resp.Body.Close()
```

Comment:
It feels like all of this code should go into a separate struct, a Zendesk client. The CDCIterator should control the CDC flow, and shouldn't care about the low level communication to the Zendesk API. We don't need to do in this PR I think.

Reply:
done